design/BufferOperations.md

```diff
@@ -0,0 +1,150 @@
+# Buffer operations
+
+This describes operations that are done on `WebGPUBuffer` objects directly and are not buffered inside a `WebGPUCommandBuffer`.
+The two primitives we need to support are the CPU writing data inside the buffer for use by the GPU (upload) and the CPU reading data produced by the GPU (readback).
+
+Design constraints are:
+
+ - For the portability of the API, prevent data races between the CPU and the GPU.
+ - For performance, minimize the number of times the data is copied around.
+ - To make the API non-blocking, only allow asynchronous readbacks.
+ - For performance on multi-process implementations, make an asynchronous upload path.
+
+## Buffer mapping
+
+### map[Write|Read] and unmap
+
+The way to have the minimal number of copies for upload and readback is to provide a buffer mapping mechanism.
+This mechanism has to be asynchronous to ensure the GPU is done using the buffer before the application can look into the ArrayBuffer.
+Otherwise on implementation where the ArrayBuffer is directly a pointer to the buffer memory, data races between the CPU and the GPU could occur.
+
+We want the status of a map operation to act as both a promise, and something that's pollable as there are advantages to both.
+`WebGPUMappedMemory` is an object that is `then`-able, meaning that it acts like a Javascript `Promise` but is pollable at the same time.
+
+The mapping operations for `WebGPUBuffer` are:
+
+```
+partial interface WebGPUBuffer {
+    WebGPUMappedMemory mapWrite(u32 offset, u32 size);
+    WebGPUMappedMemory mapRead(u32 offset, u32 size);
+};
+```
+
+These operations return new `WebGPUMappedMemory` objects representing the current range of the buffer for writing or mapping.
+The results are initialized in the "pending" state and transition at Javascript task boundary to the "available" state when the implementation can determine the GPU is done using the buffer.
+Calling `mapRead` or `mapWrite` puts the buffer in the mapped state.
+No operations are allowed in a buffer in that state except additional calls to `mapRead` or `mapWrite` and calls to `unmap`.
+In particular a mapped buffer cannot be used in a `WebGPUCommandBuffer` given to `WebGPUQueue.submit`.
+The following must be true or a validation error occurs for `mapWrite` (resp. `mapRead`):
+
+ - The buffer must have been created with the `WebGPUBufferUsage.MAP_WRITE` (resp. `WebGPUBufferUsage.MAP_READ`) usage.
+ - `offset + size` must not overflow and be at most the size of the buffer
+ - Depending on the design of memory barriers, the buffer must be, or allowed to be in the `WebGPUBufferUsage.MAP_WRITE` (resp. `WebGPUBufferUsage.MAP_READ`) usage.
+
+Then a mapped buffer can be unmapped with:
+
+```
+partial interface WebGPUBuffer {
+    void unmap();
```

Comment:
Double checked, this correctly appears in the IDL or am I missing something?

Our original idea had both `WebGPUBuffer` and `WebGPUMappedMemory` have `unmap` we thought it would be simpler to only have the `WebGPUBuffer` one in this proposal but you are right, `WebGPUMappedMemory.unmap` would allow removing a single mapping which is useful if we don't allow overlapping mappings.

Reply:
I don't think WebGPUMappedMemory.unmap is needed if we do allow overlapping mappings.